### PR TITLE
Create MITx Online daily course engagement mart and upstream models changes

### PR DESCRIPTION
**Pull request by rachellougee**
# What are the relevant tickets?

https://github.com/mitodl/hq/issues/3120 and additional intermediate models in [this spreadsheet](https://docs.google.com/spreadsheets/d/1_4z3KWh3n0XGDEAKA8DrIeKfxvUSNeX8-CVDibN7Q5A/edit#gid=1756599030)

# Description (What does it do?)

 - creating `int__mitxonline__user_courseactivity_discussion` and `int__mitxonline__user_courseactivity_daily` for the mart and other reports we need
 - Updating `int__mitxonline__user_courseactivity_video` to include additional [video events](https://edx.readthedocs.io/projects/devdata/en/latest/internal_data_formats/tracking_logs/student_event_types.html#video-interaction-events) like `complete_video`, `seek_video` etc that are useful for reports
 - creating `mart__mitxonline_course_engagement_daily` that aggregates learners course activity events daily

# How can this be tested?

Update `host` to  `mitol-ol-data-lake-production.trino.galaxy.starburst.io` under dev_production target in your profile.yml to be able to run large model

then run
```
dbt build +marts__mitxonline_course_engagements_daily  --target dev_production
```

### CODE REVIEW

File: src/ol_dbt/models/marts/mitxonline/marts__mitxonline_weekly_course_engagements.sql [hunk 1]
```diff
@@ -0,0 +1,128 @@
+{% set week_start_date= '2022-01-02' %}
```
[the remark this answers is not in the document]
Comment (rachellougee) [reply src/ol_dbt/models/marts/mitxonline/marts__mitxonline_weekly_course_engagements.sql]: Discussed with Kate, we decide to create daily course engagement mart that is unique on username+courserun_readable_id+date . We can then aggregate weekly on Preset if needed

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml [hunk 2]
```diff
@@ -1383,37 +1383,56 @@ models:
     tests:
     - not_null
   - name: useractivity_event_type
-    description: str, type of video event triggered - play_video - When a user selects
-      the video player’s play control pause_video - When a user selects the video
-      player’s pause control stop_video - When the video player reaches the end of
-      the video file and play automatically stops
+    description: str, type of video event triggered. It can be browser-emitted events
+      like play_video, pause_video, stop_video, complete_video, or mobile-emitted
+      events edx.video.played, edx.video.paused, etc.
```
Comment (pdpinch) [inline src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1388]: Would it be appropriate, and easier, to link to the open edX documentation for these events?

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml [hunk 3]
```diff
@@ -1383,37 +1383,56 @@ models:
...
     tests:
     - not_null
+  - name: useractivity_path
+    description: str, URL that generated this event
```
Comment (pdpinch) [inline src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1392]: Is this a URL or a path?

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml [hunk 4]
```diff
@@ -1383,37 +1383,56 @@ models:
...
   - name: useractivity_video_id
-    description: str, hash code for the video being watched. This value is part of
-      the courseware_studentmodule.module_id
+    description: str, hash code for the video being watched. This value is last part
```
Comment (pdpinch) [inline src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1394]:
```suggestion
    description: str, hash code for the video being watched. This value is the last part
```

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml [hunk 5]
```diff
@@ -1468,6 +1489,124 @@ models:
...
+      The value is server for this event.
+    tests:
+    - not_null
+  - name: useractivity_event_type
+    description: str, edx.grades.problem.submitted - when a problem is submitted and
+      successfully saved
+    tests:
+    - not_null
+  - name: useractivity_path
+    description: str, URL that generated this event
```
Comment (pdpinch) [inline src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml]: URL or path?

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml [hunk 6]
```diff
@@ -1468,6 +1489,124 @@ models:
...
+    description: int, open edX user ID extracted from context field
+    tests:
+    - not_null
+  - name: courserun_readable_id
+    description: str, Open edX Course ID formatted as course-v1:{org}+{course code}+{run_tag}
+    tests:
+    - not_null
+  - name: useractivity_event_source
+    description: str, specifies the source of the interaction that triggered the event.
+      The values are either browser or mobile for this event
```
Comment (pdpinch) [inline src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1560]: Is this browser or _mobile app_? Can we note that mitxonline doesn't have a mobile app at this time (but edx.org does)

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml (hunk 2 above)
Comment (rachellougee) [reply src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1388]: The documentation link https://edx.readthedocs.io/projects/devdata/en/latest/internal_data_formats/tracking_logs/student_event_types.html#video-interaction-events is too long to add in description here as pre-commit would try to format it to new line and that might not be readable. I also noticed that not all video events are documented there. e.g. complete_video

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml (hunk 3 above)
Comment (rachellougee) [reply src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml]: This field is the relative url of the page. I realized that for video events, `useractivity_page_url` is more useful here, so I updated that. But not every event has useractivity_page_url field populated.

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml [hunk 7]
```diff
@@ -1468,6 +1489,124 @@ models:
...
+    tests:
+    - not_null
+  - name: useractivity_event_type
+    description: str, type of discussion forum event triggered. Values can be edx.forum.comment.created,
+      edx.forum.comment.edited, edx.forum.response.created, edx.forum.response.voted,
+      edx.forum.thread.created, edx.forum.thread.viewed, edx.forum.thread.voted, edx.forum.searched,
+      etc.
```
Comment (pdpinch) [inline src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1567]: Can we link to open edX docs for these events?

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml (hunk 4 above)
Comment (rachellougee) [reply src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml]: updated

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml [hunk 8]
```diff
@@ -1468,6 +1489,124 @@ models:
...
+    tests:
+    - not_null
+  - name: useractivity_path
+    description: str, URL that generated this event
+  - name: useractivity_discussion_post_id
+    description: str, unique identifier for the discussion post learner engaged in.
+      Maybe Null for edx.forum.searched event
```
Comment (pdpinch) [inline src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1572]:
```suggestion
      May be Null for edx.forum.searched event
```

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml [hunk 9]
```diff
@@ -1468,6 +1489,124 @@ models:
...
+    tests:
+    - dbt_expectations.expect_column_values_to_not_be_null:
+        row_condition: "useractivity_event_type != 'edx.forum.searched'"
+  - name: useractivity_discussion_post_title
+    description: str, title for the specific discussion post. Maybe Null for edx.forum.searched
```
Comment (pdpinch) [inline src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1577]:
```suggestion
    description: str, title for the specific discussion post. May be Null for edx.forum.searched
```

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml [hunk 10]
```diff
@@ -1468,6 +1489,124 @@ models:
...
+      event.
+  - name: useractivity_discussion_block_id
+    description: str, identifier for the specific discussion component. e.g. discussion_ps2A-tab5.
+      Value is the last part of 'coursestructure_block_id' string. Maybe Null for
```
Comment (pdpinch) [inline src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1581]:
```suggestion
      Value is the last part of 'coursestructure_block_id' string. May be Null for
```

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml [hunk 11]
```diff
@@ -1468,6 +1489,124 @@ models:
...
+    description: str, URL of the page from which the discussion thread can be viewed.
+  - name: useractivity_discussion_search_query
+    description: str, the query text searched by the learner on the course discussion
+      page. Applicable for edx.forum.searched event.
+    tests:
+    - dbt_expectations.expect_column_values_to_not_be_null:
+        row_condition: "useractivity_event_type = 'edx.forum.searched'"
+  - name: useractivity_discussion_roles
+    description: str, identifies a user's roles in the discussion post. e.g. ["Student"],
+      ["Community TA", "Student"].
```
Comment (pdpinch) [inline src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml:1597]: Do we know the complete set of possible roles? Can we link to some documentation for this?

**Review by pdpinch — CHANGES_REQUESTED** [review]
Sorry my review came in bits and pieces.

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml (hunk 6 above)
Comment (rachellougee) [reply src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml]: The discussion event can only be from server, so I updated the description here. I noted mitxonline doesn't have a mobile app in video event https://github.com/mitodl/ol-data-platform/pull/947/files#diff-ed9000c0ae953ef078e3e7c9b3f50c4d2d244836cd1728a92a51e3a855be902fR1382 as that's emitted from client side.

**Comment by rachellougee** [comment]
@pdpinch Thanks for the review. I've updated description per your comments

File: src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml (hunk 2 above)
Comment (rachellougee) [reply src/ol_dbt/models/intermediate/mitxonline/_int_mitxonline__models.yml]: I added link anyway.

**Review by KatelynGit — APPROVED** [review]
I ran the dbt build and there were no errors. I also tested the table for things like blanks and to make sure we weren't duplicating records or dropping records from the source tables when creating it and it all looked good. I think this design works much better and is more flexible than the weekly one we originally discussed I'm signing off.

**Review by pdpinch — APPROVED** [review]
lgtm

**Pull request merged**